### ISSUE TITLE
“--system-reserved” kubelet option cannot work as expected

### ISSUE THREAD

**Issue by kulong0105** [kubernetes/kubernetes#72881]
**What happened**:

start kubelet with below config:
```
[root@app ~]# cat /etc/systemd/system/kubelet.service.d/10-kubeadm.conf 
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
Environment="KUBELET_SYSTEM_PODS_ARGS=--pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true"
Environment="KUBELET_NETWORK_ARGS=--network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin"
Environment="KUBELET_DNS_ARGS=--cluster-dns=10.96.0.10 --cluster-domain=cluster.local"
Environment="KUBELET_AUTHZ_ARGS=--authorization-mode=Webhook --client-ca-file=/etc/kubernetes/pki/ca.crt"
Environment="KUBELET_CADVISOR_ARGS=--cadvisor-port=8686"
Environment="KUBELET_CGROUP_ARGS=--cgroup-driver=cgroupfs"
Environment="KUBELET_CERTIFICATE_ARGS=--rotate-certificates=true --cert-dir=/var/lib/kubelet/pki"
ExecStart=/usr/bin/kubelet 
$KUBELET_KUBECONFIG_ARGS $KUBELET_SYSTEM_PODS_ARGS $KUBELET_NETWORK_ARGS $KUBELET_DNS_ARGS $KUBELET_AUTHZ_ARGS $KUBELET_CADVISOR_ARGS $KUBELET_CGROUP_ARGS $KUBELET_CERTIFICATE_ARGS $KUBELET_EXTRA_ARGS 
--docker-root=/SkyDiscovery/docker --enforce-node-allocatable=pods,kube-reserved,system-reserved 
--kube-reserved-cgroup=/kube.slice --kubelet-cgroups=/kube.slice 
--runtime-cgroups=/kube.slice --system-reserved-cgroup=/system.slice 
--kube-reserved=cpu=2,memory=1Gi --system-reserved=cpu=2,memory=1Gi 
--eviction-hard=memory.available<5%
[root@app ~]# 
```
check the related cgroup file, get the following output:
```
[root@app ~]# cat /sys/fs/cgroup/cpu/system.slice/cpu.shares 
2048
[root@app ~]# cat /sys/fs/cgroup/cpu/system.slice/cpu.cfs_period_us 
100000
[root@app ~]# cat /sys/fs/cgroup/cpu/system.slice/cpu.cfs_quota_us 
-1
[root@app ~]# cat /sys/fs/cgroup/memory/system.slice/memory.limit_in_bytes
1073741824
[root@app ~]# 
```

**What you expected to happen**:

`/sys/fs/cgroup/memory/system.slice/memory.limit_in_bytes` is not limited.

due to `memory.limit_in_bytes` is limited, may cause system services are killed by OOM, expect this config only affect the available memory resource for node(check with `kubectl describe node $node_name`)

`cpu.cfs_quota_us=-1` means not set limit for CPU usage,  but `memory.limit_in_bytes=1073741824`  means set limit for memory usage, why the function of reserved cpu and reserved memory is differnent?  

**How to reproduce it (as minimally and precisely as possible)**:

start kubelet service with config:
```
[root@app ~]# systemctl cat kubelet 
# /etc/systemd/system/kubelet.service
[Unit]
Description=kubelet: The Kubernetes Node Agent
Documentation=http://kubernetes.io/docs/
After=cephfs-mount.service
Requires=cephfs-mount.service

[Service]
ExecStartPre=/usr/bin/mkdir -p /sys/fs/cgroup/hugetlb/system.slice
ExecStartPre=/usr/bin/mkdir -p /sys/fs/cgroup/cpuset/system.slice
ExecStart=/usr/bin/kubelet
Restart=always
StartLimitInterval=0
RestartSec=10

[Install]
WantedBy=multi-user.target

# /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
Environment="KUBELET_SYSTEM_PODS_ARGS=--pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true"
Environment="KUBELET_NETWORK_ARGS=--network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin"
Environment="KUBELET_DNS_ARGS=--cluster-dns=10.96.0.10 --cluster-domain=cluster.local"
Environment="KUBELET_AUTHZ_ARGS=--authorization-mode=Webhook --client-ca-file=/etc/kubernetes/pki/ca.crt"
Environment="KUBELET_CADVISOR_ARGS=--cadvisor-port=8686"
Environment="KUBELET_CGROUP_ARGS=--cgroup-driver=cgroupfs"
Environment="KUBELET_CERTIFICATE_ARGS=--rotate-certificates=true --cert-dir=/var/lib/kubelet/pki"
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_SYSTEM_PODS_ARGS $KUBELET_NETWORK_ARGS $KUBELET_DNS_ARGS $KUBELET_AUTHZ_ARGS $KUBELET_CADVISOR_ARGS $KUBELET_CGROUP_ARGS $KUBELET_CERTIFICATE_ARGS $KUBELET_EXTRA_ARGS --docker-root=/SkyDiscovery/docker --enforce-node-allocatable=pods,kube-reserved,system-reserved --kube-reserved-cgroup=/kube.slice --kubelet-cgroups=/kube.slice --runtime-cgroups=/kube.slice --system-reserved-cgroup=/system.slice --kube-reserved=cpu=2,memory=1Gi --system-reserved=cpu=2,memory=1Gi --eviction-hard=memory.available<5%
```

**Anything else we need to know?**:

i have read this doc: https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/
and get these info:
```
Be extra careful while enforcing system-reserved reservation since it can lead to critical system
services being CPU starved or OOM killed on the node.The recommendation is to enforce 
system-reserved only if a user has profiled their nodes exhaustively to come up with precise
estimates and is confident in their ability to recover if any process in that group is oom_killed.
```
so it is reasonable to limit the file  `/sys/fs/cgroup/memory/system.slice/memory.limit_in_bytes` ? 

if not use `--system-reserved`, the ` /sys/fs/cgroup/cpu/system.slice/cpu.shares ` will be `1024` by default, which will result in system services being CPU starved

**Environment**:
- Kubernetes version (use `kubectl version`):   v1.10.12
- Cloud provider or hardware configuration:  dell server PowerEdge R740
- OS (e.g. from /etc/os-release):  CentOS7.4
- Kernel (e.g. `uname -a`):  3.10.0-693.el7.centos.x86.64
- Install tools: 
- Others:

**Comment by JishanXing** [comment]
> Be extra careful while enforcing system-reserved reservation since it can lead to critical system
> services being CPU starved or OOM killed on the node.The recommendation is to enforce 
> system-reserved only if a user has profiled their nodes exhaustively to come up with precise
> estimates and is confident in their ability to recover if any process in that group is oom_killed.
> 
> so it is reasonable to limit the file /sys/fs/cgroup/memory/system.slice/memory.limit_in_bytes ?
I think this is the design behavior to limit the system services when user specify system-reserved.

**Comment by Imr0219** [comment]
```
...
ExecStart=/usr/bin/kubelet \
$KUBELET_KUBECONFIG_ARGS $KUBELET_SYSTEM_PODS_ARGS $KUBELET_NETWORK_ARGS $KUBELET_DNS_ARGS $KUBELET_AUTHZ_ARGS $KUBELET_CADVISOR_ARGS $KUBELET_CGROUP_ARGS $KUBELET_CERTIFICATE_ARGS $KUBELET_EXTRA_ARGS 
--docker-root=/SkyDiscovery/docker --enforce-node-allocatable=pods,kube-reserved,system-reserved \
--kube-reserved-cgroup=/kube.slice --kubelet-cgroups=/kube.slice \
--runtime-cgroups=/kube.slice --system-reserved-cgroup=/system.slice \
--kube-reserved=cpu=2,memory=1Gi --system-reserved=cpu=2,memory=1Gi \
--eviction-hard=memory.available<5%
```
This can work well

**Comment by sjenning** [comment]
/reopen

xref https://bugzilla.redhat.com/show_bug.cgi?id=1866794#c5

I think this is indeed an issue.

We should be setting `cpu.shares` on the `system-reserved-cgroup`, otherwise there is no way to ensure that the cpu part of the system reservation is actually accessible to the system daemons when under contention.  This problem gets worse the larger the reservation and throws of the whole calculation of shares (relative) to cpu millicores (absolute).

@derekwaynecarr @dashpole what do you think?

**Comment by sjenning** [comment]
This could get tricky though.
https://github.com/kubernetes/kubernetes/blob/7740b8124c2f684de3caeae0f2cc5d2a1329d43e/pkg/kubelet/cm/node_container_manager_linux.go#L63-L130
kubelet does not currently assume/default `SystemReservedCgroup` and doesn't require that the user provide it unless `system-reserved` is included in the `EnforceNodeAllocatable` keys.  Thus we would have to default `SystemReservedCgroup` to something based on the `CgroupDriver` (`/system.slice` for `systemd` for example). Making `SystemReservedCgroup` required when `EnforceNodeAllocatable` does not included `system-reserved` is backward incompatible.

**Comment by dashpole** [comment]
Extending my thoughts from sig-node.

1. If I set EnforceNodeAllocatable on `pods` (the default), then that requires adjusting shares for all `/kubepods` peers to be correct.  I think it is a bug if we don't do this.
2. If I set --kube-reserved or --system-reserved (and the corresponding cgroups flag), and have `EnforceNodeAllocatable=pods` we should _at least_ set cpu.shares on the peer cgroup of /kubepods that contains --kube-reserved-cgroup or system-reserved-cgroup.  This follows from (1) because if we are already setting shares on the cgroup, we can be smart enough to set it correctly.

I think even just those two are sufficient to make our behavior sensible, even if it isn't completely correct.  If we agree on those, there are some follow-up questions we could answer.

If we don't set EnforceNodeAllocatable _at all_, should we set shares?  I'm thinking we shouldn't, as this allows "turning off" all enforcement.

There is still the point of wanting to enforce CPU shares without memory limits...

**Comment by derekwaynecarr** [comment]
@dashpole I agree with both points.  I think disabling EnforceNodeAlloctable is in effect disabling resource management at this point in Kubernetes lifecycle.  It is all best-effort at that point.

**Comment by povsister** [comment]
I think this is designed behavior.

By default, kubelet only enforces `pods` allocatable. It sets `cpu.shares=<nodeCapacity> - <systemReserved> - <kubeReserved> - <hardEvictionThreshold>` and `memory.limit_in_bytes=<same as cpu.shares>` for `<cgroupRoot>/kubepods` (cgroupfs driver) or `<cgroupRoot>/kubepods.slice` (systemd driver).

Let's have an example:

Config kubelet as following. We reserve some resources for kube and system, but not enforce them.
```yaml
cgroupDriver: systemd
cgroupRoot: /
enforceNodeAllocatable:
  - pods
kubeReserved:
  memory: 1Gi
  cpu: 500m # equals 0.5 cpu
kubeletCgroups: /kubereserved.slice/kubelet.service
kubeReservedCgroup: /kubereserved.slice
systemReserved:
  memory: 1.5Gi
  cpu: 500m
systemReservedCgroup: /system.slice
```

You will get the following allocatable resource from `kubectl describe node`
```yaml
Capacity:
  cpu:                8
  ephemeral-storage:  92115000Ki
  hugepages-2Mi:      0
  memory:             16266012Ki  # 15.5125 GiB
  pods:               110
Allocatable:
  cpu:                7 # Apparently 7 = 8 - 0.5 - 0.5
  ephemeral-storage:  84893183860
  hugepages-2Mi:      0
  memory:             13542172Ki  # 12.9148 GiB = 15.5125 GiB - 1 GiB - 1.5 GiB - 100 MiB (default hard eviction threshold)
  pods:               110
```

Let check the cgroup settings
```bash
[root@linux ~]# cat /sys/fs/cgroup/cpu/kubepods.slice/cpu.shares
7168 # 7 cpu cores
[root@linux ~]# cat /sys/fs/cgroup/cpu/kubepods.slice/cpu.cfs_quota_us
-1  # unlimited
[root@linux ~]# cat /sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes
13972041728  # = 13644572 KiB, which comes from 13542172Ki (nodeAllocatable) + 100MiB (hardEvictionThreshold)

[root@linux ~]# cat /sys/fs/cgroup/cpu/kubereserved.slice/cpu.shares
1024 # default value. Not "enforced"
[root@linux ~]# cat /sys/fs/cgroup/memory/kubereserved.slice/memory.limit_in_bytes
9223372036854771712  # default value. Not "enforced" either.
```

Conclusion:
* `systemReserved` and `kubeReserved` will always affect `nodeAllocatable`, which will be used for pod scheduling reference and eviction decision.
*  `systemReservedCgroup` and `kubeReservedCgroup` are only required and enforced if corresponding key are set in `enforceNodeAllocatable`.
* `cpu.shares` only guarantees cpu time under contention, it allows other process borrowing cpu if idle. `cpu.cfs_quota_us` instead limits the maximum cpu usage, even there are lots of cpu in idle. I think setting `cpu.shares` is enough for "reservation".
* If you "enforce" `systemReserved` or `kubeReserved`, it's not surprising that a `cpu.shares` and a `memory.limit_in_bytes` will be enforced on corresponding cgroups.

By the way, I might found a bug when kubelet is handling `systemReservedCgroup` and `kubeReservedCgroup` , that why I am here 🤣

**Comment by fgiloux** [comment]
@povsister I agree with most of what you wrote except that:
 [root@linux ~]# cat /sys/fs/cgroup/cpu/kubepods.slice/cpu.shares
7168 # 7 cpu cores

These are not 7 cpu cores. They are shares. In your example the node has 8 cores. How much you get out of these 8 cores under contention is depending on the configuration of the other slices at the same level. Let say you have 1024 in system.slice and 1024 in users.slice the total would be 9x1024. you get 7/9 of 8 cores for pods:  6.22 cores and 1.77 is left for the other 2 slices.

Now let say that you have not 8 but 96 cores. You keep the same value for kube-reserved and system-reserved. This gives:
/kubepods.slice/cpu.shares 95x1024
The other 2 slices are unchanged: 1024
You then have 95/97 of 96 cores for pods: ~94 cores and ~2 cores are left for the other 2 slices (twice what you thought).

The thing is that with big machines these 2 cores won't still be enough and you try to reserve more cores for system processes, let say 6 in total.
/kubepods.slice/cpu.shares 90x1024
The other 2 slices are unchanged: 1024
You then have 90/92 of 96 cores for pods, which is still close to ~94 cores and ~2 cores are left for the other 2 slices. Adding 5 cores to kube-reserved + system-reserved ended up adding ~ 0.1 core to what the non-pod processes get under contention. Changing that would really help. Hopefully my explanation was not too confusing. I tried to reflect in calculations what  kulong0105 and dashpole wrote.

**Comment by povsister** [comment]
> @povsister I agree with most of what you wrote except that:
> [root@linux ~]# cat /sys/fs/cgroup/cpu/kubepods.slice/cpu.shares
> 7168 # 7 cpu cores

Thanks for reminding me with nice examples. I know exactly what you want to explain: They are shares calculated in relative percentage of the same level.
Sorry I just forgot that when I am rushing to submit another issue #101189 

Looks like there are still lots of room to improve current implemented node allocatable calculation😄

**Comment by fromanirh** [comment]
per https://github.com/kubernetes/kubernetes/issues/72881#issuecomment-821022253 and https://github.com/kubernetes/kubernetes/issues/72881#issuecomment-821224980 it seems this is the designed behaviour, perhaps we should improve the docs but it should be handled on a separate issue. Please reopen if we get new information.
/close

**Comment by fgiloux** [comment]
@fromanirh I can't reopen the issue but frankly I don't follow your reasoning. The comments you took as justification for "work as designed" clearly show that the calculation is not right, which is nicely summarized in:
"Looks like there are still lots of room to improve current implemented node allocatable calculation"

**Comment by fromanirh** [comment]
> @fromanirh I can't reopen the issue but frankly I don't follow your reasoning. The comments you took as justification for "work as designed" clearly show that the calculation is not right, which is nicely summarized in:
> "Looks like there are still lots of room to improve current implemented node allocatable calculation"

Fair enough, I may have miss a key detail. How we can improve the allocatable calculation? We would like actionable items to make progress.

**Comment by fromanirh** [comment]
ok, answering myself I missed the implications of the proposal of dashpole here: https://github.com/kubernetes/kubernetes/issues/72881#issuecomment-672154398 . This is the improvement which we should do to address this issue.
/triage accepted
/area kubelet
/priority important-longterm

**Comment by danielfoehrKn** [comment]
> You then have 90/92 of 96 cores for pods, which is still close to ~94 cores and ~2 cores are left for the other 2 slices. Adding 5 cores to kube-reserved + system-reserved ended up adding ~ 0.1 core to what the non-pod processes get under contention.

What @fgiloux correctly pointed out here is that users THINK they reserved 5 CPU cores, but in fact here they only reserved 0.1 cores.
I think that's pretty important & I think a bunch of people have misconfigured their kube/system-reserved settings because of that.

On the same page, always assigning 1024 shares for peer groups of `kubepods`, does not allow to lower the minimally guaranteed CPU time for e.g system.slice as the kubepods `cpu.shares` are capped at 1024 * `number of cores`

**Comment by umialpha** [comment]
it seems to be bug to me.

**Comment by harche** [comment]
WIP PR - https://github.com/kubernetes/kubernetes/pull/125982

**Comment by natasha41575** [comment]
@harche is this still an issue? I see the PR is merged

**Comment by harche** [comment]
With https://github.com/kubernetes/kubernetes/pull/125982 merged, this can be closed.
